Rubberduck.Core/UI/CodeExplorer/CodeExplorerAddComponentService.cs

```diff
@@ -0,0 +1,122 @@
+
+            _parseManager.OnSuspendParser(
+                this,
+                Enum.GetValues(typeof(ParserState)).Cast<ParserState>(),
+                () => _addComponentService.AddComponent(projectId, componentType, code, prefixInModule));
+        }
+
+        public void AddComponentWithAttributes(CodeExplorerItemViewModel node, ComponentType componentType, string code, string additionalPrefixInModule = null)
+        {
+            var projectId = ProjectId(node);
+            if (projectId == null)
+            {
+                return;
+            }
+
+            var folderAnnotation = FolderAnnotation(node);
+            var optionCompare = OptionCompareStatement();
+
+            var basePrefixInModule = optionCompare != null
+                ? $"{folderAnnotation}{Environment.NewLine}{optionCompare}"
+                : folderAnnotation;
+
+            var prefixInModule = additionalPrefixInModule != null
+                ? $"{basePrefixInModule}{Environment.NewLine}{additionalPrefixInModule}"
+                : basePrefixInModule;
+
+            _parseManager.OnSuspendParser(
+                this,
+                Enum.GetValues(typeof(ParserState)).Cast<ParserState>(),
+                () => _addComponentService.AddComponentWithAttributes(projectId, componentType, code, prefixInModule));
+        }
+
+        private string ProjectId(CodeExplorerItemViewModel node)
+        {
+            return node?.Declaration.ProjectId;
+        }
+
+        private string FolderAnnotation(CodeExplorerItemViewModel node)
+        {
+            return (node is CodeExplorerCustomFolderViewModel folder) 
+                ? folder.FolderAttribute 
+                : $"'@Folder(\"{Folder(node)}\")";
+        }
+
+        private string Folder(CodeExplorerItemViewModel node)
+        {
+            var declaration = node?.Declaration;
+            if (declaration == null)
+            {
+                return ActiveProjectName();
+            }
+
+            var customFolder = declaration.CustomFolder;
+            if (customFolder != null)
+            {
+                return customFolder.Replace("\"", string.Empty);
+            }
+
+            return ProjectFolder(declaration.ProjectName);
+        }
+
+        private string ActiveProjectFolder()
+        {
+            return ProjectFolder(ActiveProjectName());
+        }
+
+        private string ActiveProjectName()
+        {
+            using (var activeProject = _vbe.ActiveVBProject)
+            {
+                return activeProject?.Name;
+            }
+        }
+
+        private static string ProjectFolder(string projectName)
+        {
+            return projectName;
+        }
+
+        private string OptionCompareStatement()
+        {
+            using (var hostApp = _vbe.HostApplication())
+            {
+                return hostApp?.ApplicationName == "Access" 
+                    ? "Option Compare Database" 
```

Comment:
Referencing #4628 -- this probably ought to be cleaned up when a PR is opened to address that issue.